File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/SumTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.beam.sdk.transforms;

import static org.apache.beam.sdk.testing.CombineFnTester.testCombineFn;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;

import org.apache.beam.sdk.coders.BigEndianIntegerCoder;
import org.apache.beam.sdk.coders.BigEndianLongCoder;
import org.apache.beam.sdk.coders.CoderRegistry;
import org.apache.beam.sdk.coders.DoubleCoder;
import org.apache.beam.sdk.coders.VarIntCoder;
import org.apache.beam.sdk.coders.VarLongCoder;
import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Lists;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/** Tests for Sum. */
@RunWith(JUnit4.class)
public class SumTest {
  private static final CoderRegistry STANDARD_REGISTRY = CoderRegistry.createDefault();

  @Test
  public void testSumGetNames() {
    assertEquals("Combine.globally(SumInteger)", Sum.integersGlobally().getName());
    assertEquals("Combine.globally(SumDouble)", Sum.doublesGlobally().getName());
    assertEquals("Combine.globally(SumLong)", Sum.longsGlobally().getName());
    assertEquals("Combine.perKey(SumInteger)", Sum.integersPerKey().getName());
    assertEquals("Combine.perKey(SumDouble)", Sum.doublesPerKey().getName());
    assertEquals("Combine.perKey(SumLong)", Sum.longsPerKey().getName());
  }

  @Test
  public void testSumIntegerFn() {
    testCombineFn(Sum.ofIntegers(), Lists.newArrayList(1, 2, 3, 4), 10);
  }

  @Test
  public void testSumLongFn() {
    testCombineFn(Sum.ofLongs(), Lists.newArrayList(1L, 2L, 3L, 4L), 10L);
  }

  @Test
  public void testSumDoubleFn() {
    testCombineFn(Sum.ofDoubles(), Lists.newArrayList(1.0, 2.0, 3.0, 4.0), 10.0);
  }

  @Test
  public void testSumDoubleFnInfinity() {
    testCombineFn(
        Sum.ofDoubles(),
        Lists.newArrayList(Double.NEGATIVE_INFINITY, 2.0, 3.0, Double.POSITIVE_INFINITY),
        Double.NaN);
  }

  @Test
  public void testSumDoubleFnPositiveInfinity() {
    testCombineFn(
        Sum.ofDoubles(),
        Lists.newArrayList(1.0, 2.0, 3.0, Double.POSITIVE_INFINITY),
        Double.POSITIVE_INFINITY);
  }

  @Test
  public void testSumDoubleFnNegativeInfinity() {
    testCombineFn(
        Sum.ofDoubles(),
        Lists.newArrayList(Double.NEGATIVE_INFINITY, 2.0, 3.0, 4.0),
        Double.NEGATIVE_INFINITY);
  }

  @Test
  public void testSumDoubleFnNan() {
    testCombineFn(Sum.ofDoubles(), Lists.newArrayList(1.0, 2.0, 3.0, Double.NaN), Double.NaN);
  }

  @Test
  public void testGetAccumulatorCoderEquals() {
    Combine.BinaryCombineIntegerFn sumIntegerFn = Sum.ofIntegers();
    assertEquals(
        sumIntegerFn.getAccumulatorCoder(STANDARD_REGISTRY, VarIntCoder.of()),
        sumIntegerFn.getAccumulatorCoder(STANDARD_REGISTRY, VarIntCoder.of()));
    assertNotEquals(
        sumIntegerFn.getAccumulatorCoder(STANDARD_REGISTRY, VarIntCoder.of()),
        sumIntegerFn.getAccumulatorCoder(STANDARD_REGISTRY, BigEndianIntegerCoder.of()));

    Combine.BinaryCombineLongFn sumLongFn = Sum.ofLongs();
    assertEquals(
        sumLongFn.getAccumulatorCoder(STANDARD_REGISTRY, VarLongCoder.of()),
        sumLongFn.getAccumulatorCoder(STANDARD_REGISTRY, VarLongCoder.of()));
    assertNotEquals(
        sumLongFn.getAccumulatorCoder(STANDARD_REGISTRY, VarLongCoder.of()),
        sumLongFn.getAccumulatorCoder(STANDARD_REGISTRY, BigEndianLongCoder.of()));

    Combine.BinaryCombineDoubleFn sumDoubleFn = Sum.ofDoubles();
    assertEquals(
        sumDoubleFn.getAccumulatorCoder(STANDARD_REGISTRY, DoubleCoder.of()),
        sumDoubleFn.getAccumulatorCoder(STANDARD_REGISTRY, DoubleCoder.of()));
  }
}

File: sdks/go/pkg/beam/transforms/stats/sum_test.go
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package stats

import (
	"testing"

	"github.com/apache/beam/sdks/v2/go/pkg/beam"
	"github.com/apache/beam/sdks/v2/go/pkg/beam/testing/passert"
	"github.com/apache/beam/sdks/v2/go/pkg/beam/testing/ptest"
)

// TestSumInt verifies that Sum adds ints correctly.
func TestSumInt(t *testing.T) {
	tests := []struct {
		in  []int
		exp []int
	}{
		{
			[]int{1, -2, 3},
			[]int{2},
		},
		{
			[]int{1, 11, 7, 5, 10},
			[]int{34},
		},
		{
			[]int{0},
			[]int{0},
		},
	}

	for _, test := range tests {
		p, s, in, exp := ptest.CreateList2(test.in, test.exp)
		passert.Equals(s, Sum(s, in), exp)

		if err := ptest.Run(p); err != nil {
			t.Errorf("Sum(%v) != %v: %v", test.in, test.exp, err)
		}
	}
}

// TestSumFloat verifies that Sum adds float32 correctly.
func TestSumFloat(t *testing.T) {
	tests := []struct {
		in  []float32
		exp []float32
	}{
		{
			[]float32{1, -2, 3.5},
			[]float32{2.5},
		},
		{
			[]float32{0, -99.99, 1, 1},
			[]float32{-97.99},
		},
		{
			[]float32{5.67890},
			[]float32{5.6789},
		},
	}

	for _, test := range tests {
		p, s, in, exp := ptest.CreateList2(test.in, test.exp)
		passert.Equals(s, Sum(s, in), exp)

		if err := ptest.Run(p); err != nil {
			t.Errorf("Sum(%v) != %v: %v", test.in, test.exp, err)
		}
	}
}

// TestSumKeyed verifies that Sum works correctly for KV values.
func TestSumKeyed(t *testing.T) {
	tests := []struct {
		in  []student
		exp []student
	}{
		{
			[]student{{"alpha", 1}, {"beta", 4}, {"charlie", 3.5}},
			[]student{{"alpha", 1}, {"beta", 4}, {"charlie", 3.5}},
		},
		{
			[]student{{"alpha", 1}},
			[]student{{"alpha", 1}},
		},
		{
			[]student{{"alpha", 1}, {"alpha", -4}, {"beta", 4}, {"charlie", 0}, {"charlie", 5.5}},
			[]student{{"alpha", -3}, {"beta", 4}, {"charlie", 5.5}},
		},
	}

	for _, test := range tests {
		p, s, in, exp := ptest.CreateList2(test.in, test.exp)
		kv := beam.ParDo(s, studentToKV, in)
		sum := SumPerKey(s, kv)
		sumStudent := beam.ParDo(s, kvToStudent, sum)
		passert.Equals(s, sumStudent, exp)

		if err := ptest.Run(p); err != nil {
			t.Errorf("Sum(%v) != %v: %v", test.in, test.exp, err)
		}
	}
}
